File: 930-binary-subarrays-with-sum/930-binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, A: List[int], S: int) -> int:
        def atMost(S):
            if S < 0:
                return 0
            preSum = res = l = 0
            for r in range(len(A)):
                preSum += A[r]
                while preSum > S:
                    preSum -= A[l]
                    l += 1
                res += r - l + 1
            return res
        return atMost(S) - atMost(S - 1)

File: 930-binary-subarrays-with-sum/930-binary-subarrays-with-sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        return atMostK(nums, goal) - atMostK(nums, goal-1);
    }
    
    int atMostK(vector<int> &array, int K){
        if(K < 0){
            return 0;
        }
        int size = array.size();
        int result = 0;
        int left = 0;
        int right = 0;
        int preSum = 0;
        
        for(int right = 0; right < size; ++right){
            preSum += array[right];
            while(preSum > K){
                preSum -= array[left];
                ++left;
            }
            int lengthOfSubarray = right - left + 1;
            result += lengthOfSubarray;
        }
        
        return result;
    }
};